etcd: extend rate limiting to consider the number of inflight requests

giorio94 opened the pull request:
Reporting the description of the last commit for convenience:

```
Currently, the etcd rate limiter operates based on a token bucket of a
fixed size. Yet, this is problematic when etcd is overloaded, since we
may continue to issue new requests even if it cannot keep up. Hence,
let's start using the custom APILimiter, which also takes into account
the number of currently inflight requests. The maximum amount of
inflight requests can configured through the etcd.maxInflight parameter,
and defaults to the same value as etcd.qps if unset.
```

<!-- Description of change -->

```release-note
etcd: extend rate limiting to consider the number of inflight requests
```

giorio94 committed:
etcd: add max inflight requests to rate limiter

Currently, the etcd rate limiter operates based on a token bucket of a
fixed size. Yet, this is problematic when etcd is overloaded, since we
may continue to issue new requests even if it cannot keep up. Hence,
let's start using the custom APILimiter, which also takes into account
the number of currently inflight requests. The maximum amount of
inflight requests can configured through the etcd.maxInflight parameter,
and defaults to the same value as etcd.qps if unset.

The rate limiter wait is removed when checking the status of the etcd
endpoints, to prevent that the check fails due to rate limiting, as that
would put a high pressure on etcd in case the connection gets restarted.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
daemon: extract APILimiterObserver implementation to separate package

This commit extracts the APILimiterObserver implementation from the
daemon package to a separate one, to allow for its reuse.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1055,9 +1094,22 @@ reList:
+			case <-ctx.Done():
+				return
+			default:
+				goto recreateWatcher
```
We are missing `rl.Done()` here, right?

giorio94 replied:
We don't need to (and cannot) call `rl.Done()` when `Wait()` returns an error, since `rl` is nil in that case.

marseel replied:
ah yea, I don't know why I thought error was coming from different place...

oblazek commented on pkg/kvstore/etcd.go:
```diff
@@ -1162,11 +1214,16 @@ func (e *etcdClient) determineEndpointStatus(ctx context.Context, endpointAddres
 	e.getLogger().Debugf("Checking status to etcd endpoint %s", endpointAddress)
 
-	e.limiter.Wait(ctxTimeout)
+	rl, err := e.limiter.Wait(ctxTimeout)
```
just a small nit: is there a reason you are mixing lr/rl ? e.g. `rl, err := e.limiter.Wait(ctxTimeout)` and `lr, err := e.limiter.Wait(ctx)`

giorio94 replied:
Thanks for noticing, they should be consistent now.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -1162,7 +1214,6 @@ func (e *etcdClient) determineEndpointStatus(ctx context.Context, endpointAddres
 
 	e.getLogger().Debugf("Checking status to etcd endpoint %s", endpointAddress)
 
-	e.limiter.Wait(ctxTimeout)
```
I've dropped the rate limiter wait here, to prevent that the status check fails due to rate limiting, as that would put a high pressure on etcd in case the connection gets restarted.

giorio94 committed:
CODEOWNERS: assign pkg/rate/metrics to @cilium/metrics

Let's assign the newly introduced `pkg/rate/metrics` folder to
the @cilium/metrics team.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1002,8 +1037,13 @@ reList:
+			continue
+		}
 		kvs, revision, err := e.paginatedList(ctx, scopedLog, w.Prefix)
+		// Using lr.Error for convenience, as it matches lr.Done() when err is nil
```
err is always nil here though, right?

giorio94 replied:
Not necessarily, `e.paginatedList` might return an error if one of the Get operations failed.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1270,7 +1322,12 @@ func (e *etcdClient) GetIfLocked(ctx context.Context, key string, lock KVLocker)
-	e.limiter.Wait(ctx)
+	lr, err := e.limiter.Wait(ctx)
+	if err != nil {
+		increaseMetric(key, metricRead, "GetLocked", duration.EndError(err).Total(), err)
```
Can we do `defer increaseMetric(key, metricRead, "GetLocked", duration.EndError(err).Total(), err)` two lines above (not inside if) and remove another increaseMetric below?

giorio94 replied:
We cannot with the current `increaseMetric` implementation, since `duration.EndError(err).Total()` would be executed immediately, hence not counting correctly the duration. It should work modifying `increaseMetric` to take the span as input (rather than the pre-computed duration), but the change tends to be on the large side. WDYT?

marseel replied:
ah yea, we could do that in separate PR.

giorio94 replied:
I'm giving it a try. If it ends up being too large I'll open a separate PR

giorio94 replied:
I've updated the PR adding one more commit which does the `increaseMetrics` deferring, to avoid having to repeat that in case of errors. @marseel PTAL

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1295,12 +1354,19 @@ func (e *etcdClient) Get(ctx context.Context, key string) (bv []byte, err error)
-	e.limiter.Wait(ctx)
+	lr, err := e.limiter.Wait(ctx)
+	if err != nil {
+		increaseMetric(key, metricRead, "Get", duration.EndError(err).Total(), err)
```
Similar defer here.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1315,7 +1381,12 @@ func (e *etcdClient) GetPrefixIfLocked(ctx context.Context, prefix string, lock
-	e.limiter.Wait(ctx)
+	lr, err := e.limiter.Wait(ctx)
+	if err != nil {
+		increaseMetric(prefix, metricRead, "GetPrefixLocked", duration.EndError(err).Total(), err)
```
Similar defer here.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1341,12 +1415,19 @@ func (e *etcdClient) GetPrefix(ctx context.Context, prefix string) (k string, bv
-	e.limiter.Wait(ctx)
+	lr, err := e.limiter.Wait(ctx)
+	if err != nil {
+		increaseMetric(prefix, metricRead, "GetPrefix", duration.EndError(err).Total(), err)
```
Similar defer here.

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -1359,7 +1440,14 @@ func (e *etcdClient) Set(ctx context.Context, key string, value []byte) (err err
-	e.limiter.Wait(ctx)
+	lr, err := e.limiter.Wait(ctx)
+	if err != nil {
+		increaseMetric(key, metricSet, "Set", duration.EndError(err).Total(), err)
```
Similar defer here and bunch of below :)

marseel reviewed (approved):
lgtm except for small nits.

giorio94 committed:
kvstore: refactor metrics increase for etcd methods

This commit slightly refactors the approach adopted to increase the
kvstore metrics, deferring this operation at the beginning of each
function (similarly to how tracing is handled). This allows to
transparently handle early returns due to errors.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
